### PR TITLE
Fix build system

### PR DESCRIPTION
**Pull request by mubarak23**

### CODE REVIEW

File: build_system/src/build.rs [hunk 1]
```diff
@@ -29,6 +29,16 @@ impl BuildArg {
                         );
                     }
                 }
+                "--sysroot" => {
+                    if let Some(arg) = args.next() {
+                        build_arg.flags.push("--sysroot".to_string());
```
Comment (antoyo) [inline build_system/src/build.rs]: This is not what should be done.

Instead, this new `--sysroot` flag that you added should be used to determine if we should build the sysroot or not. This happens [here](https://github.com/rust-lang/rustc_codegen_gcc/blob/master/build_system/src/build.rs#L223).

File: build_system/src/test.rs [hunk 2]
```diff
@@ -138,6 +138,16 @@ impl TestArg {
                         return Err("Expected an argument after `--features`, found nothing".into())
                     }
                 },
+                "--sysroot" => match args.next() {
```
Comment (antoyo) [inline build_system/src/test.rs]: This file already contain an option to build the sysroot [here](https://github.com/antoyo/rustc_codegen_gcc/blob/master/build_system/src/test.rs#L30), so it should not be needed.

File: build_system/src/build.rs (hunk 1 above)
Comment (mubarak23) [reply build_system/src/build.rs]: instead we should call `build_sysroot(&env, &args.config_info)?;` on the if check

File: build_system/src/test.rs (hunk 2 above)
Comment (mubarak23) [reply build_system/src/test.rs]: i will remove this

File: build_system/src/build.rs (hunk 1 above)
Comment (antoyo) [reply build_system/src/build.rs]: In this function, set a new field that you can add on this struct to `true` and in the link from my comment above, wrap the call to `build_sysroot` in an `if` that will check if this new field is set to `true`.

Comment (mubarak23) [reply build_system/src/build.rs]: adding a boolean flag on this `BuildArg`, then set it to TRUE on `--sysroot` passed

File: build_system/src/build.rs [hunk 3]
```diff
@@ -9,6 +9,7 @@ use std::path::Path;
 struct BuildArg {
     flags: Vec<String>,
     config_info: ConfigInfo,
+    sys_root: bool,
```
Comment (antoyo) [inline build_system/src/build.rs]: I would prefer this spelling:
```suggestion
    sysroot: bool,
```

**Review by antoyo — COMMENTED** [review]
The CI failed, so the code needs a fix, but looks good otherwise.

**Comment by mubarak23** [comment]
what could make the CI failed

**Comment by antoyo** [comment]
You can see the results of the CI at the bottom of this page. If you click on [this one, for example](https://github.com/antoyo/rustc_codegen_gcc/actions/runs/8410641421/job/23029333783?pr=9#step:10:1), it shows you the build errors.

**Comment by antoyo** [comment]
There's still one error (and a warning) to fix in the CI. Keep up the good work.

Btw, do you get notified for the CI failures?

**Comment by mubarak23** [comment]
Yes I get notifications for CI failure ,
I will look into it and push a fix

**Comment by mubarak23** [comment]
the error is from  `arg.sysroot = true;` on looking at the code again, i am suppose to use `build_arg.sysroot= true`

**Comment by antoyo** [comment]
The new error in the CI is because you now need to use this new `--sysroot` flag in the CI jobs [that are in this directory](https://github.com/antoyo/rustc_codegen_gcc/tree/master/.github/workflows).

**Comment by mubarak23** [comment]
you mean something like this `cargo clippy --all-targets --sysroot master -- -D warnings` in ci.yml file

**Comment by antoyo** [comment]
No, the argument you added is for the `build` command from the build system, so whenever you see `./y.sh build`, please add `--sysroot` after.

**Comment by mubarak23** [comment]
Alright

File: build_system/src/build.rs [hunk 4]
```diff
@@ -220,8 +225,11 @@ fn build_codegen(args: &mut BuildArg) -> Result<(), String> {
     })?;
 
     println!("[BUILD] sysroot");
```
Comment (antoyo) [inline build_system/src/build.rs]: Please move this line inside the `if`.

**Review by antoyo — COMMENTED** [review]
It seems the sysroot is still not built in the CI.

Does this work locally on your side?

File: build_system/src/build.rs (hunk 4 above)
Comment (mubarak23) [reply build_system/src/build.rs]: Alright, could be the reason why the CI is failing

**Comment by mubarak23** [comment]
> It seems the sysroot is still not built in the CI.
> 
> Does this work locally on your side?

am getting the following on local `macbookpro@MacBooks-MBP-2 rustc_codegen_gcc % ./y.sh build --sysroot`
```
[BUILD] build system
    Finished `release` profile [optimized] target(s) in 0.02s
```

File: build_system/src/build.rs [hunk 5]
```diff
@@ -224,7 +224,6 @@ fn build_codegen(args: &mut BuildArg) -> Result<(), String> {
         )
     })?;
 
-    println!("[BUILD] sysroot");
     if args.sysroot {
```
Comment (antoyo) [inline build_system/src/build.rs:222]: I mean to actually move it like this:
```
     println!("[BUILD] sysroot");		
     if args.sysroot {
```

**Comment by antoyo** [comment]
> > It seems the sysroot is still not built in the CI.
> > Does this work locally on your side?
> 
> am getting the following on local `macbookpro@MacBooks-MBP-2 rustc_codegen_gcc % ./y.sh build --sysroot`
> 
> ```
> [BUILD] build system
>     Finished `release` profile [optimized] target(s) in 0.02s
> ```

You're missing the actual build of the sysroot.
If you run `./y.sh build` on the master branch, you get this which is the expected result:
```
[BUILD] build system
   Compiling y v0.1.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 1.35s
    Finished `dev` profile [optimized + debuginfo] target(s) in 0.05s
[BUILD] sysroot
    Updating crates.io index
   Compiling core v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/core)
   Compiling compiler_builtins v0.1.108
   Compiling libc v0.2.153
   Compiling std v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/std)
   Compiling rustc-std-workspace-core v1.99.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/rustc-std-workspace-core)
   Compiling alloc v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/alloc)
   Compiling cfg-if v1.0.0
   Compiling adler v1.0.2
   Compiling memchr v2.7.1
   Compiling rustc-demangle v0.1.23
   Compiling unwind v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/unwind)
   Compiling rustc-std-workspace-alloc v1.99.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/rustc-std-workspace-alloc)
   Compiling panic_unwind v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/panic_unwind)
   Compiling panic_abort v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/panic_abort)
   Compiling gimli v0.28.1
   Compiling miniz_oxide v0.7.2
   Compiling std_detect v0.1.5 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/stdarch/crates/std_detect)
   Compiling object v0.32.2
   Compiling hashbrown v0.14.3
   Compiling addr2line v0.21.0
   Compiling rustc-std-workspace-std v1.99.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/rustc-std-workspace-std)
   Compiling proc_macro v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/proc_macro)
   Compiling unicode-width v0.1.11
   Compiling getopts v0.2.21
   Compiling test v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/test)
   Compiling sysroot v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 48.27s
```
Do you know why it doesn't work with the code from this PR?

File: build_system/src/build.rs (hunk 5 above)
Comment (mubarak23) [reply build_system/src/build.rs:222]: i see

**Comment by mubarak23** [comment]
> > > It seems the sysroot is still not built in the CI.
> > > Does this work locally on your side?
> > 
> > 
> > am getting the following on local `macbookpro@MacBooks-MBP-2 rustc_codegen_gcc % ./y.sh build --sysroot`
> > ```
> > [BUILD] build system
> >     Finished `release` profile [optimized] target(s) in 0.02s
> > ```
> 
> You're missing the actual build of the sysroot. If you run `./y.sh build` on the master branch, you get this which is the expected result:
> 
> ```
> [BUILD] build system
>    Compiling y v0.1.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_system)
>     Finished `release` profile [optimized] target(s) in 1.35s
>     Finished `dev` profile [optimized + debuginfo] target(s) in 0.05s
> [BUILD] sysroot
>     Updating crates.io index
>    Compiling core v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/core)
>    Compiling compiler_builtins v0.1.108
>    Compiling libc v0.2.153
>    Compiling std v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/std)
>    Compiling rustc-std-workspace-core v1.99.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/rustc-std-workspace-core)
>    Compiling alloc v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/alloc)
>    Compiling cfg-if v1.0.0
>    Compiling adler v1.0.2
>    Compiling memchr v2.7.1
>    Compiling rustc-demangle v0.1.23
>    Compiling unwind v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/unwind)
>    Compiling rustc-std-workspace-alloc v1.99.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/rustc-std-workspace-alloc)
>    Compiling panic_unwind v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/panic_unwind)
>    Compiling panic_abort v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/panic_abort)
>    Compiling gimli v0.28.1
>    Compiling miniz_oxide v0.7.2
>    Compiling std_detect v0.1.5 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/stdarch/crates/std_detect)
>    Compiling object v0.32.2
>    Compiling hashbrown v0.14.3
>    Compiling addr2line v0.21.0
>    Compiling rustc-std-workspace-std v1.99.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/rustc-std-workspace-std)
>    Compiling proc_macro v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/proc_macro)
>    Compiling unicode-width v0.1.11
>    Compiling getopts v0.2.21
>    Compiling test v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot/sysroot_src/library/test)
>    Compiling sysroot v0.0.0 (/home/bouanto/Ordinateur/Programmation/Rust/Projets/rustc_codegen_gcc/build_sysroot)
>     Finished `dev` profile [unoptimized + debuginfo] target(s) in 48.27s
> ```
> 
> Do you know why it doesn't work with the code from this PR?

i have no idea, i will switch to the master and run the build

**Comment by mubarak23** [comment]
Looking at logs again, i keep seeing error below,  [as found here](https://github.com/antoyo/rustc_codegen_gcc/actions/runs/8412062927/job/23032447253?pr=9)

```
 ---- lang_tests::gep.rs stderr ----

error[E0463]: can't find crate for `std` 
```

```
error: could not compile `rustc_codegen_gcc` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: linking with `cc` failed: exit status: 1
``` [found here](https://github.com/antoyo/rustc_codegen_gcc/actions/runs/8412062922/job/23032446290?pr=9)

**Comment by antoyo** [comment]
Yes, this is because the sysroot is not actually built.
Somehow, your PR changed the build system in a way that prevent the sysroot to be built (even with the `--sysroot` flag).
You need to figure out why this is happening.
As mentioned before, you should see `[BUILD] sysroot` in the logs.

**Comment by mubarak23** [comment]
I will look at the places where I made the changes on my branch and compare it the place on the master  branch

**Comment by mubarak23** [comment]
after looking at where i made change on my PR Branch and the master branch, this is piece of code that i added`create_dir(gccjit_target)?;`

File: build_system/src/build.rs [hunk 6]
```diff
@@ -204,11 +209,20 @@ fn build_codegen(args: &mut BuildArg) -> Result<(), String> {
...
-
-    println!("[BUILD] sysroot");
-    build_sysroot(&env, &args.config_info)?;
+    fs::create_dir_all(gccjit_target).map_err(|error| {
+        format!(
+            "Failed to create directory `{}`: {:?}",
+            gccjit_target, error
+        )
+    })?;
+
```
Comment (antoyo) [inline build_system/src/build.rs:220]: Perhas you could try print the field to see if we actually reach this line before the `if`:
```
println!("Build sysroot: {}", args.sysroot);
```

Comment (MubarakCinderbuild) [reply build_system/src/build.rs:220]: Yaah, i will try this

Comment (antoyo) [reply build_system/src/build.rs:220]: It is not even printed, so we don't even reach this code.
Do you have any idea what in this PR could have the effect of skipping this code?

Comment (mubarak23) [reply build_system/src/build.rs:220]: I suspect the build command

Comment (antoyo) [reply build_system/src/build.rs:220]: You can try adding `println!` calls in a few places (especially where you changed the code) to see where it stops happening.

Comment (mubarak23) [reply build_system/src/build.rs:220]: so far, this is the only place, we added the sysroot flag   
```
- name: Build
      run: |
        ./y.sh prepare --only-libcore
        ./y.sh build --sysroot
        cargo test
```

in the ci.yml file
i remove it from all other places i added it

**Comment by mubarak23** [comment]
i will add `args.sysroot = false; `in run() function

**Comment by mubarak23** [comment]
where could this error come from `error[E0463]: can't find crate for `std``

**Comment by antoyo** [comment]
This is because the sysroot is not built. Building the sysroot means building `std` among other things.

**Comment by mubarak23** [comment]
should i pull out a new branch from master and push my work on it, i feel like the current branch on this PR is currupt

**Comment by antoyo** [comment]
Yes, you can do that.

**Comment by antoyo** [comment]
Was superceeded by #13.

**Pull request closed without merging**